### PR TITLE
[stable/jenkins] Update and standardize Jenkins chart

### PR DESCRIPTION
**Pull request by binoculars**
## Changes
* Standardize template filenames
* Use current helm best practices with values file (lowercase names, etc)
* Update default image repositories to their current locations
* Use current LTS master and latest agent image tags
* Compatible with Kubernetes 1.6+, if rbac is enabled (alpha API is deprecated)

/cc @icereval

### CODE REVIEW

File: stable/jenkins/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
+{{- $component := printf "%s-master" .Release.Name }}
```
Comment (unguiculus) [inline stable/jenkins/templates/deployment.yaml]: In `NOTES.txt` you do this: `"component={{ template "fullname" . }}-master"`. Please make it consistent. Actually, I would not even add the fullname to the component name. I'd just go like this:

```yaml
name: {{ template "fullname" . }}-master
labels:
  ...
  component: master
```

File: stable/jenkins/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  labels:
+    app: {{ template "name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+    component: {{ $component }}
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
```
Comment (unguiculus) [inline stable/jenkins/templates/deployment.yaml]: Add `app` and `release` labels to selector.

File: stable/jenkins/templates/networkpolicy.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{{- if .Values.networkPolicy.enabled }}
+kind: NetworkPolicy
+apiVersion: {{ .Values.networkPolicy.apiVersion }}
+metadata:
+  name: {{ .Release.Name }}-master
```
Comment (unguiculus) [inline stable/jenkins/templates/networkpolicy.yaml]: Add standard labels.

File: stable/jenkins/templates/networkpolicy.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+spec:
+  podSelector:
+    matchLabels:
+      component: {{ .Release.Name }}-master
```
Comment (unguiculus) [inline stable/jenkins/templates/networkpolicy.yaml]: Add `app` and `release` labels to selector.

File: stable/jenkins/templates/networkpolicy.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      - port: {{ .Values.master.service.externalPort }}
+    # Allow inbound connections from slave
+    - from:
+      - podSelector:
+          matchLabels:
+            "jenkins/{{ .Release.Name }}-agent": "true"
+      ports:
+      - port: {{ .Values.agentlistenerService.externalPort }}
+{{- if .Values.agent.enabled }}
+---
```
Comment (unguiculus) [inline stable/jenkins/templates/networkpolicy.yaml]: One file per resource, please.

File: stable/jenkins/templates/service.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  - port: {{ .Values.service.externalPort }}
+    targetPort: {{ .Values.service.internalPort }}
+    protocol: TCP
+    name: {{ .Values.service.name }}
+  selector:
+    app: {{ template "name" . }}
+    release: {{ .Release.Name }}
+    component: {{ .Release.Name }}-{{ .Values.service.name }}
+{{- if .Values.agent.enabled }}
+---
```
Comment (unguiculus) [inline stable/jenkins/templates/service.yaml]: One file per resource, please.

File: stable/jenkins/templates/networkpolicy.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+kind: NetworkPolicy
+apiVersion: {{ .Values.networkPolicy.apiVersion }}
+metadata:
```
Comment (unguiculus) [inline stable/jenkins/templates/networkpolicy.yaml]: Add standard labels.

File: stable/jenkins/templates/pvc.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  labels:
+    app: {{ template "fullname" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+  annotations:
+  {{- if .Values.persistence.storageClass }}
+    volume.beta.kubernetes.io/storage-class: {{ .Values.persistence.storageClass | quote }}
+  {{- else }}
+    volume.alpha.kubernetes.io/storage-class: default
```
Comment (unguiculus) [inline stable/jenkins/templates/pvc.yaml]: There's no use in supporting the alpha version any more. Here's an interesting approach for supporting storage classes. I'd suggest you follow it:
https://github.com/kubernetes/charts/pull/1332

File: stable/jenkins/templates/serviceaccount.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+{{- if .Values.rbac.install }}
```
Comment (unguiculus) [inline stable/jenkins/templates/serviceaccount.yaml]: @mgoodness has come up with a nice approach for RBAC. We've started suggesting following his example. I'd be great if you did too. See:

https://github.com/kubernetes/charts/pull/1504/files
https://github.com/mgoodness/charts/tree/master/stable/kube2iam

File: stable/jenkins/templates/test.yaml [hunk 10]
```diff
@@ -0,0 +1,48 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: "{{ .Release.Name }}-ui-test-{{ randAlphaNum 5 | lower }}"
```
Comment (unguiculus) [inline stable/jenkins/templates/test.yaml:4]: Add standard labels.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Excellent. This is very much appreciated. I made a few comments.

**Comment by binoculars** [comment]
@unguiculus I made the requested changes. I'm still unsure if the ClusterRole is correct. I would appreciate any help there and a good look at the NetworkPolicies.

**Comment by unguiculus** [comment]
@mgoodness Could you review the RBAC stuff?

**Comment by binoculars** [comment]
Let me know when you'll be ready to review and I'll rebase it, but feel free to make any changes otherwise.

**Comment by unguiculus** [comment]
Waiting for the rebase actually. Tests won't run as long as there are conflicts.

File: stable/jenkins/README.md [hunk 11]
```diff
@@ -91,61 +79,93 @@ Agent:
 
 The suported volume types are: `ConfigMap`, `EmptyDir`, `HostPath`, `Nfs`, `Pod`, `Secret`. Each type supports a different set of configurable attributes, defined by [the corresponding Java class](https://github.com/jenkinsci/kubernetes-plugin/tree/master/src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes).
```
Comment (evansj) [inline stable/jenkins/README.md]: *supported

**Comment by unguiculus** [comment]
@binoculars Please fix conflicts.

**Comment by binoculars** [comment]
@unguiculus Conflicts resolved. I'll be around to make changes if necessary.

File: stable/jenkins/templates/service.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  ports:
+    - port: {{ .Values.service.externalPort }}
+      targetPort: {{ .Values.service.internalPort }}
+      protocol: TCP
+      name: http
+  selector:
+    app: {{ template "name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
```
Comment (unguiculus) [inline stable/jenkins/templates/service.yaml]: Remove `heritage` and `chart` from selector.

File: stable/jenkins/templates/deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    release: {{ .Release.Name }}
+    component: {{ .Values.master.name }}
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: {{ template "name" . }}
+        chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+        heritage: {{ .Release.Service }}
```
Comment (unguiculus) [inline stable/jenkins/templates/deployment.yaml]: Remove `chart` and `heritage` from selector.

File: stable/jenkins/templates/service.yaml [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  {{- end }} 
+  ports:
+    - port: {{ .Values.service.externalPort }}
+      targetPort: {{ .Values.service.internalPort }}
+      protocol: TCP
+      name: http
+  selector:
+    app: {{ template "name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    heritage: {{ .Release.Service }}
```
Comment (unguiculus) [inline stable/jenkins/templates/service.yaml]: Remove `chart` and `heritage` from selector.

File: stable/jenkins/templates/test.yaml (hunk 10 above)
Comment (unguiculus) [reply stable/jenkins/templates/test.yaml:4]: Standard labels still missing.

File: stable/jenkins/templates/test-configmap.yaml [hunk 15]
```diff
@@ -5,5 +5,5 @@ metadata:
 data:
   run.sh: |-
     @test "Testing Jenkins UI is accessible" {
-      curl --retry 12 --retry-delay 10 {{ .Release.Name }}-jenkins:8080{{ default "" .Values.Master.JenkinsUriPrefix }}/login
+      curl --retry 12 --retry-delay 10 {{ .Release.Name }}-jenkins:8080{{ default "" .Values.master.uriPrefix }}/login
```
Comment (unguiculus) [inline stable/jenkins/templates/test-configmap.yaml]: Add an empty default in `values.yaml`, then you don't need to specify a default here:

`{{ .Values.master.uriPrefix }}`

Comment (unguiculus) [reply stable/jenkins/templates/test-configmap.yaml]: Needs standard labels.

File: stable/jenkins/values.yaml [hunk 16]
```diff
@@ -3,86 +3,57 @@
...
+  #  -Dcom.sun.management.jmxremote.ssl=false 
+  #JENKINS_OPTS: >
+  #  --prefix=/jenkins 
+  # Optionally configure a JMX port
+  #jmxPort: 4000
+  # Node labels and tolerations for pod assignment
+  # ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
+  nodeSelector: {}
+  # ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#taints-and-tolerations-beta-feature
+  tolerations: {}
```
Comment (unguiculus) [inline stable/jenkins/values.yaml]: `tolerations: []`

File: stable/jenkins/templates/NOTES.txt [hunk 17]
```diff
@@ -1,36 +1,36 @@
...
-{{- else if contains "LoadBalancer" .Values.Master.ServiceType }}
+{{- else if contains "LoadBalancer" .Values.service.type }}
   NOTE: It may take a few minutes for the LoadBalancer IP to be available.
         You can watch the status of by running 'kubectl get svc --namespace {{ .Release.Namespace }} -w {{ template "fullname" . }}'
   export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "fullname" . }} --template "{{ "{{ range (index .status.loadBalancer.ingress 0) }}{{ . }}{{ end }}" }}")
-  echo http://$SERVICE_IP:{{ .Values.Master.ServicePort }}/login
+  echo http://$SERVICE_IP:{{ .Values.service.externalPort }}/login
 
-{{- else if contains "ClusterIP"  .Values.Master.ServiceType }}
+{{- else if contains "ClusterIP"  .Values.service.type }}
```
Comment (unguiculus) [inline stable/jenkins/templates/NOTES.txt]: Label selector wrong. Needs `app`, `release`, and `component` labels.

File: stable/jenkins/values.yaml [hunk 18]
```diff
@@ -3,86 +3,57 @@
...
-#  - |
-#    print 'adding global pipeline libraries, register properties, bootstrap jobs...'
-  CustomConfigMap: false
-# Node labels and tolerations for pod assignment
-# ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
-# ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#taints-and-tolerations-beta-feature
-  NodeSelector: {}
-  Tolerations: {}
+master:
+  name: master
```
Comment (unguiculus) [inline stable/jenkins/values.yaml]: You are still keeping the name of the master component configurable. Doesn't make sense to me. Why would you ever change this?

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by binoculars** [comment]
@unguiculus requested changes made

**Comment by lachie83** [comment]
I'm happy with this change. We should just bump the major version as this will be backwards incompatible. How do we want to handle that? Also can you please take a look at why CI is failing

**Comment by unguiculus** [comment]
Please fix conflicts and apply suggested changes.

**Comment by unguiculus** [comment]
@binoculars Any change you can continue with this? I'd really like to get it in.

**Comment by binoculars** [comment]
@unguiculus yes, it'll take me a bit to resolve the conflicts, but I'll do it soon.

**Comment by unguiculus** [comment]
@binoculars I have to mark this as stale now. Please update it within one week.

**Comment by unguiculus** [comment]
Closing this as stale now. Feel free to reopen when you decide to work on it again.

**Pull request closed without merging**